src/PrinterCartridgeLevelAlert.php

```diff
@@ -0,0 +1,364 @@
+<?php
+
+/**
+ * ---------------------------------------------------------------------
+ *
+ * GLPI - Gestionnaire Libre de Parc Informatique
+ *
+ * http://glpi-project.org
+ *
+ * @copyright 2015-2023 Teclib' and contributors.
+ * @copyright 2003-2014 by the INDEPNET Development Team.
+ * @licence   https://www.gnu.org/licenses/gpl-3.0.html
+ *
+ * ---------------------------------------------------------------------
+ *
+ * LICENSE
+ *
+ * This file is part of GLPI.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+/**
+ *  PrinterCartridgeLevelAlert class
+ *
+ * @ since version 0.84
+ **/
+class PrinterCartridgeLevelAlert extends CommonGLPI
+{
+    protected static $notable = true;
+
+    public static function getTypeName($nb = 0)
+    {
+        return __('Printer Cartridge Level Alerts');
+    }
+
+    /**
+     * @since 0.85
+     **/
+    public static function canView()
+    {
+        return (Session::haveRight('printer', READ) && Session::haveRight('cartridge', READ));
+    }
+
+    /**
+     * Print a good title
+     *
+     *@return void
+     **/
+    public static function title()
+    {
+        Html::displayTitle(
+            "",
+            self::getTypeName(),
+            "<i class='fas fa-check fa-lg me-2'></i>" . self::getTypeName()
+        );
+    }
+
+    /**
+    * @param $entities
+    *        $repeat for already notified management
+    *
+    * @return string
+    */
+    private static function query($entities, $repeat = null)
+    {
+        global $DB;
+
+        $where = "c.date_out is NULL AND ( ( l.value REGEXP '^[0-9]+$' AND l.value <= i.warn_level ) OR ( l.value like 'WARNING' OR l.value like 'BAD' )) AND l.printers_id = p.id AND p.entities_id IN ($entities)";
```

Comment:
Also, using `LIKE` with a simple text value does not make sense, it's easier to simply use equals

Reply:
I need to check it attentively. Will do it tomorrow. BTW, I wonder if "l.printers_id = p.id" is really usefull. I get the same results without it, but it could be related to my data set... When could the entity be empty? Is it just to prevent errors in the future?

Need also to check again, but last time I had some troubles with this kind of expressions:   ['l.value' => ['<=' => 'i.warn_level']] as the parser considered that the 'i.warn_level' was a string and not a table field...

Reply:
> Operators notations like `['LIKE' => 'BAD']` is wrong; correct syntax is `['LIKE', 'BAD']` (see https://glpi-developer-documentation.readthedocs.io/en/master/devapi/database/dbiterator.html#operators)

I was thinking it was intentional for a case insensitive check, but it would be best to avoid the possibility of the case being different between records.

Reply:
Seems like equals is case insensitive as well

Reply:
> Seems like equals is case insensitive as well

The case insensivity comes from the DB itself by default. There are particular treatments to do if you want to work with case-sensitive operations.

Reply:
Guys,

using  ```'p.entities_id' => $entities``` leads to ```AND `p`.`entities_id` = ''19', '24', '22', '25', '23'' AND ``` which is not correct from MariaDB POV.

Reply:
BTW, ```AND (`l`.`value` IN ('i.warn_level'))) OR``` is what I was talking about yesterday evening...

Reply:
> Guys,
> 
> using `'p.entities_id' => $entities` leads to `` AND `p`.`entities_id` = ''19', '24', '22', '25', '23'' AND `` which is not correct from MariaDB POV.

For sur, because you call `$query  = self::query($_SESSION["glpiactiveentities_string"]);`; which is... a sstring, and not an array.

> 
> `'p.entities_id' => [$entities]` could be better if... in `` p`.`entities_id` IN (''19', '24', '22', '25', '23'') `` the constructor was not putting each entity id between quotes...

Reply:
> > using `'p.entities_id' => $entities` leads to `` AND `p`.`entities_id` = ''19', '24', '22', '25', '23'' AND `` which is not correct from MariaDB POV.
> 
> For sure, because you call `$query = self::query($_SESSION["glpiactiveentities_string"]);`; which is... a sstring, and not an array.

OK, my fault, I'll change it to ```$query  = self::query($_SESSION["glpiactiveentities"]); ```

Reply:
> BTW, `` AND (`l`.`value` IN ('i.warn_level'))) OR `` is what I was talking about yesterday evening...

But this is still blocking...

Reply:
OK, with some small changes it seems to be correct now. Launching tests...